pkg/apis/core/types.go

```diff
@@ -4767,3 +4773,47 @@ const (
 	// DefaultHardPodAffinityWeight defines the weight of the implicit PreferredDuringScheduling affinity rule.
 	DefaultHardPodAffinitySymmetricWeight int32 = 1
 )
+
+type UnsatisfiableConstraintResponse string
+
+const (
+	// DoNotSchedule instructs the scheduler not to schedule the pod
+	// when constraints are not satisfied.
+	DoNotSchedule UnsatisfiableConstraintResponse = "DoNotSchedule"
+	// ScheduleAnyway instructs the scheduler to schedule the pod
+	// even if constraints are not satisfied.
+	ScheduleAnyway UnsatisfiableConstraintResponse = "ScheduleAnyway"
+)
+
+// TopologySpreadConstraint specifies how to spread matching pods among the given topology.
+type TopologySpreadConstraint struct {
+	// MaxSkew describes the degree of imbalance of pods distribution.
+	// It's the max difference between the number of matching pods in any two
+	// topology domains against a given topology type.
+	// For example, in a 3-zone cluster, currently pods with the same labelSelector
+	// spread as 1/1/0:
+	// - if MaxSkew is 1, incoming pod can only be scheduled to zone3 to become 1/1/1;
+	// schedule it onto zone1(zone2) will make the ActualSkew(2) violates MaxSkew(1)
+	// - if MaxSkew is 2, incoming pod can be scheduled to any zone.
+	// It's a required value. Default value is 1 and 0 is not allowed.
+	MaxSkew int32
```

Comment:
should we make it unsigned ?

Reply:
I also thought of this `uint32`. But it seems we're not enforcing a rule - `DeploymentSpec.Replicas` is using `int32`.

Reply:
any suggestions from @kubernetes/sig-api-machinery-api-reviews  ?